server/container/util/src/main/java/org/apache/james/util/CloseableIterator.java

```diff
@@ -0,0 +1,106 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+import org.apache.james.util.streams.Iterators;
+
+public interface CloseableIterator<T> extends Iterator<T>, Closeable {
```

Comment:
> why is it an interface?

Because I have already 2 implementations of it:
 - `ScrollIterator` which clears the scroll context when required
 - And `CloseableIterator.Impl` which is used to cary along the Closeable stuff in the `ScrollIterator -> Stream -> Iterator` transformation.

Globaly, ScrollIterator needs to access the last request, in order to get the scrollId and clear it. You can not implement this feature by leveraging `CloseableIterator.Impl`

More generally, an iterator that I can close is a rather abstract and generic things so I don't see why we should not have an interface describing that behaviour.

Reply:
I think that `Stream` is close enough to your need

Reply:
Let's give it a try even if that mean changing Manager APIs...